lib/check-meta.nix

```diff
@@ -0,0 +1,121 @@
+  # e.g brokenness or license.
+  #
+  # Return { valid: Bool } and additionally
+  # { reason: String; errormsg: String } if it is not valid, where
+  # reason is one of "unfree", "blacklisted" or "broken".
+  checkMetaValidity = { meta, config, system }: let
+
+    whitelist = config.whitelistedLicenses or [];
+    blacklist = config.blacklistedLicenses or [];
+
+    onlyLicenses = list:
+      lib.lists.all (license:
+        let l = lib.licenses.${license.shortName or "BROKEN"} or false; in
+        if license == l then true else
+          throw ''‘${showLicense license}’ is not an attribute of lib.licenses''
+      ) list;
+
+    areLicenseListsValid =
+      if lib.mutuallyExclusive whitelist blacklist then
+        assert onlyLicenses whitelist; assert onlyLicenses blacklist; true
+      else
+        throw "whitelistedLicenses and blacklistedLicenses are not mutually exclusive.";
+
+    hasWhitelistedLicense = assert areLicenseListsValid; attrs:
+      meta ? license && builtins.elem meta.license whitelist;
+
+    hasBlacklistedLicense = assert areLicenseListsValid; attrs:
+      meta ? license && builtins.elem meta.license blacklist;
+
+    isUnfree = licenses: lib.lists.any (l:
+      !l.free or true || l == "unfree" || l == "unfree-redistributable") licenses;
+
+    allowUnfree = config.allowUnfree or false || builtins.getEnv "NIXPKGS_ALLOW_UNFREE" == "1";
+
+    # Alow granular checks to allow only some unfree packages
+    # Example:
+    # {pkgs, ...}:
+    # {
+    #   allowUnfree = false;
+    #   allowUnfreePredicate = (x: pkgs.lib.hasPrefix "flashplayer-" x.name);
+    # }
+    allowUnfreePredicate = config.allowUnfreePredicate or (x: false);
+
+    # Check whether unfree packages are allowed and if not, whether the
+    # package has an unfree license and is not explicitely allowed by the
+    # `allowUNfreePredicate` function.
+    hasDeniedUnfreeLicense = meta:
+      !allowUnfree &&
+      meta ? license &&
+      isUnfree (lib.lists.toList meta.license) &&
+      !allowUnfreePredicate meta;
+
+    showLicense = license: license.shortName or "unknown";
+
+    allowBroken = config.allowBroken or false || builtins.getEnv "NIXPKGS_ALLOW_BROKEN" == "1";
+
+    in # compose the result
+      if hasDeniedUnfreeLicense meta && !(hasWhitelistedLicense meta) then
+        { valid = false; reason = "unfree"; errormsg = "has an unfree license (‘${showLicense meta.license}’)"; }
+      else if hasBlacklistedLicense meta then
+        { valid = false; reason = "blacklisted"; errormsg = "has a blacklisted license (‘${showLicense meta.license}’)"; }
+      else if !allowBroken && meta.broken or false then
+        { valid = false; reason = "broken"; errormsg = "is marked as broken"; }
+      else if !allowBroken && meta.platforms or null != null && !lib.lists.elem system meta.platforms then
+        { valid = false; reason = "broken"; errormsg = "is not supported on ‘${system}’"; }
+      else { valid = true; };
+  # ^ checkMetaValidity
```

Comment:
In any case, this discussion thread is on a stale diff.  Current PR is: https://github.com/NixOS/nixpkgs/pull/27045